# Week 1 Genome Assembly -- Feedback

1 + 0.75 + 0 + 0 + 0 + 1 + 1 + 0.33 + 0 + 0 = 4.08 points out of 10 possible points

1. Question 1.1, 1.4 how many reads (0.5 pts each)

  * good --> +1

2. Question 1.2, 1.4 simulation script(s)

  * good code, but it only is set up for the first question with 5x coverage. You can use sys.argv to pass the number of reads/coverage as an input argument or copy/paste your code and in the second code block increase the number of reads for 15x coverage --> +0.75

3. Question 1.2, 1.4 plotting script(s)

  * I don't see code for plotting --> +0

4. Question 1.2, 1.4 histograms with overlaid Poisson distributions (0.5 pts each)

  * I don't see the plots; please add --> +0

5. Question 1.3, 1.4 how much of genome not sequenced/comparison to Poisson expectations (0.5 pts each, 0.25 for answer and 0.25 for code)

  * don't see answers or code for this --> +0

6. Question 2 De novo assembly (0.5 pts each, 0.25 for answer and 0.25 for code)

  * number of contigs --> +0.5
  * total length of contigs --> +0.5

7. Question 2 De novo assembly cont (0.5 pts each, 0.25 for answer and 0.25 for code)

  * size of largest contig --> +0.5
  * contig n50 size --> +0.5

8. whole genome alignment (0.33 pts each, 0.33/2 for answer and 0.33/2 for code)

  * average identity --> +0.33
  * length of longest alignment --> don't see this answer + 0
  * how many insertions and deletions in assembly --> don't see this answer + 0

9. decoding the insertion (0.5 pts each, 0.25 for answer and 0.25 for code)

  * position of insertion in assembly --> don't see this answer + 0
  * length of novel insertion --> don't see this answer + 0

10. decoding the insertion cont (0.5 pts each, 0.25 for answer and 0.25 for code)

  * DNA sequence of encoded message --> don't see this answer + 0
  * secret message --> don't see this answer + 0
